bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java

```diff
@@ -0,0 +1,221 @@
+                    // no language root
+                    try {
+                        if (relationshipManager.isSource(resource)) {
+                            // the resource is a blueprint
+                            RangeIterator liveCopiesIterator = relationshipManager.getLiveRelationships(resource, null, null);
+                            if (liveCopiesIterator != null) {
+                                LiveRelationship relationship = (LiveRelationship) liveCopiesIterator.next();
+                                LiveCopy liveCopy = relationship.getLiveCopy();
+                                if (liveCopy != null) {
+                                    root = liveCopy.getBlueprintPath();
+                                }
+                            }
+                        } else if (relationshipManager.hasLiveRelationship(resource)) {
+                            // the resource is a live copy
+                            LiveRelationship liveRelationship = relationshipManager.getLiveRelationship(resource, false);
+                            if (liveRelationship != null) {
+                                LiveCopy liveCopy = liveRelationship.getLiveCopy();
+                                if (liveCopy != null) {
+                                    root = liveCopy.getPath();
+                                }
+                            }
+                        }
+                    } catch (WCMException e) {
+                        // ignore
+                    }
+                }
+            }
+        }
+        return root;
+    }
+
+    /**
+     * Returns the localization root of the experience fragment path based on the localization root of the current page.
+     *
+     * As of today (08/aug/2019) the XF UI does not support creating Live and Language Copies, which prevents getRoot
+     * to be used with XF.
+     * This method works around this issue by deducting the XF root from the XF path and the root of the current page.
+     *
+     * @param xfPath The experience fragment path.
+     * @param currentPageRoot The localization root of the current page.
+     * @return The localization root of the experience fragment path if it exists, {@code null} otherwise.
+     */
+    private String getXfRoot(String xfPath, String currentPageRoot) {
+        String xfRoot = null;
+        if (!StringUtils.isEmpty(xfPath) && !StringUtils.isEmpty(currentPageRoot)
+            && resolver.getResource(xfPath) != null && resolver.getResource(currentPageRoot) != null) {
+            String[] xfPathTokens = Text.explode(xfPath, PATH_DELIMITER_CHAR);
+            String[] referenceRootTokens = Text.explode(currentPageRoot, PATH_DELIMITER_CHAR);
+            int xfRootDepth = referenceRootTokens.length + 1;
+            if (xfPathTokens.length >= xfRootDepth) {
+                String[] xfRootTokens = new String[xfRootDepth];
+                System.arraycopy(xfPathTokens, 0, xfRootTokens, 0, xfRootDepth);
```

Comment:
as alternative solution you could use `Arrays` class e.g.
`String[] xfRootTokens = Arrays.copyOf(xfPathTokens, xfRootDepth);`
or
`String[] xfRootTokens = Arrays.copyOfRange(xfPathTokens, 0, xfRootDepth);`
in my opinion the above approaches/methods are more descriptive and more intuitive

Reply:
The way it is now might be a bit faster as `Arrays.copyOf`relies internally on `System.arraycopy`.
See more comments in https://stackoverflow.com/questions/2589741/what-is-more-efficient-system-arraycopy-or-arrays-copyof